src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/TdsParserStateObject.Multiplexer.cs

```diff
@@ -0,0 +1,427 @@
+
+                        ReadOnlySpan<byte> payloadSource = data.Slice(0, payloadBytesAvailable);
+                        Span<byte> payloadTarget = partialPacket.Buffer.AsSpan(partialPacket.CurrentLength, payloadBytesAvailable);
+                        payloadSource.CopyTo(payloadTarget);
+
+                        partialPacket.CurrentLength = partialPacket.CurrentLength + payloadBytesAvailable;
+                        bytesConsumed += payloadBytesAvailable;
+                        data = data.Slice(payloadBytesAvailable);
+                    }
+                    else if (partialPacket.CurrentLength > partialPacket.RequiredLength)
+                    {
+                        // the partial packet contains a complete packet of data and also contains
+                        // data from a following packet
+
+                        // the TDS spec requires that all packets be of the defined packet size apart from
+                        // the last packet of a response. This means that is is not possible to have more than
+                        // 2 packet fragments in a single buffer like this:
+                        //  - first packet caused the partial
+                        //  - second packet is the one we have just unpacked
+                        //  - third packet is the extra data we have found
+
+                        // we must throw an exception because we have encountered an invalid tds stream
+                        throw SQL.ParsingError(ParsingErrorState.CorruptedTdsStream);
+                    }
+
+                    if (partialPacket.CurrentLength == partialPacket.RequiredLength)
+                    {
+                        // partial packet has been completed
+                        consumePartialPacket = true;
+                    }
+                }
+
+                if (bytesConsumed > 0)
+                {
+                    if (data.Length > 0)
+                    {
+                        // some data has been taken from the buffer, put into the partial
+                        // packet buffer and we have data left so move the data we have
+                        // left to the start of the buffer so we can pass the buffer back
+                        // as zero based to the caller avoiding offset calculations everywhere
+                        Buffer.BlockCopy(
+                            dataBuffer, dataOffset + bytesConsumed, // from
+                            dataBuffer, dataOffset, // to
+                            dataLength - bytesConsumed // for
+                        );
+#if DEBUG
+                        // for debugging purposes fill the removed data area with an easily
+                        // recognisable pattern so we can see if it is misused
+                        Span<byte> removed = dataBuffer.AsSpan(dataOffset + (dataLength - bytesConsumed), (dataOffset + bytesConsumed));
```

Comment:
Is it "removing" too much data here? should it be:
`Span<byte> removed = dataBuffer.AsSpan(dataOffset + (dataLength - bytesConsumed), bytesConsumed);`
instead of
`Span<byte> removed = dataBuffer.AsSpan(dataOffset + (dataLength - bytesConsumed), (dataOffset + bytesConsumed))`;

(i.e. `dataOffset + bytesConsumed` is the index of the end marker, but this should "just" be the length - which is just `bytesConsumed`, right?)

Reply:
I think you're right. If I make that change no tests fail but that doesn't mean it's ok just that I don't have coverage on that case. Change it and see if any behaviour changes.

Reply:
This path is covered by the ReconstrucMultiplePacketSequenceWithShortEnd test if you want to test it.

Reply:
[SQLMultiplexor.zip](https://github.com/user-attachments/files/17440708/SQLMultiplexor.zip) the standalone test harness.

Reply:
Thanks. I don't think the general code is the issue here, as when the Cancellation is removed from the test, it keeps reading the data fine. So it must be something that happens when the request is cancelled